---
title: "01_peaks"
author: "kurt"
date: "4/24/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
library(IRanges)
source("../util/my_class_functions.R")

setwd("/scratch/Shares/rinnclass/CLASS_2023/kurt/Final_Project/01_peaks")

# create folders
if (dir.exists("results") == F){
  dir.create("results")}

if (dir.exists("results/gene_annotations") == F){
  dir.create("results/gene_annotations")}

if (dir.exists("analysis") == F){
  dir.create("analysis")}
  
if (dir.exists("analysis/consensus_peaks") == F){  
  dir.create("analysis/consensus_peaks")}

if (dir.exists("figures") == F){
  dir.create("figures")}

### Paths
basepath <- "/scratch/Shares/rinnclass/CLASS_2023/kurt/Final_Project"

# filepath to import peaks
broadpeakfilepath <- "/scratch/Shares/rinnclass/CLASS_2023/data/data/peaks"
```

```{r import large files}
# import peaks
if (exists("peak_list") == F){
peak_list <- import_peaks(consensus_file_path = broadpeakfilepath)}


# gencode_gr -> gene -> protein coding / lncRNA
if (exists("gencode_gr") == F){
gencode_gr <- rtracklayer::import("/scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/gencode.v32.annotation.gtf")}

```

```{r consensus peaks}

### create consensus peaks
# find unique dpbs
dbps <- unique(sapply(names(peak_list), function(x) {
   unlist(strsplit(x, "_"))[1]
}))

# now run our function consensus_from_reduced to get consensus list for each dbp
consensus_list <- lapply(dbps, consensus_from_reduced, peak_list)
names(consensus_list) <- dbps

# number of consensus peaks for each dbp
num_consensus_peaks <- sapply(consensus_list, length) %>% 
  as.data.frame() %>%
  rownames_to_column( var = "dbp") %>%
  dplyr::rename(number_consensus_peaks = ".")

# only keep dbps with >= 1000 peaks
filtered_consensus_list <- consensus_list[sapply(consensus_list, length) >= 1000]

### save/export files
write_csv(peak_num, "results/num_peaks_df.csv")
save(consensus_list, file = "results/consensus_list.RData")
save(consensus_list, file = "results/filtered_consensus_list.RData")

#export filtered consensus peak files as .bed
consensus_path <- "01_peaks/analysis/consensus_peaks/"
exportpath <- file.path(basepath, consensus_path)
for(i in 1:length(filtered_consensus_list)) {
rtracklayer::export(consensus_list[[i]], paste0(exportpath, names(consensus_list)[i], "_consensus_peaks.bed") )}
```
```{r gene annotations}
### annotations
# genes
gencode_genes <- gencode_gr[gencode_gr$type == "gene"] 
mrna_genes <- gencode_genes[gencode_genes$gene_type %in% "protein_coding"] 
lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% "lncRNA"]
mrna_lncrna_genes <- gencode_genes[gencode_genes$gene_type %in% c("protein_coding","lncRNA")]

# gene ids
mrna_gene_ids <-mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "protein_coding"]
lncrna_gene_ids <- mrna_lncrna_genes$gene_id[mrna_lncrna_genes$gene_type == "lncRNA"]

# promoters
lncrna_mrna_promoters <- promoters(mrna_lncrna_genes, upstream = 1000, downstream = 1000)
```

```{r peaks overlapping gene promoters and gene bodies}
# How many peaks overlap gene annotations ??

# number of peaks each dbp has
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_list),
                           "num_peaks" = sapply(filtered_consensus_list, length))


# Now let's get the total amount of the genome covered by all the peaks for a given DBP.
num_peaks_df$total_peak_length <- sapply(filtered_consensus_list, function(x) sum(width(x)))

# counting promoter overlaps
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, type = "counts")

# ok we see this is a large matrix of the number of overlaps at each promoter
# cols are promoters, rows are DBPS
# so now we just row sum for each DBP !

num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)

# Now let's break these promoters into two groups "lncrna" and "mrna"
# We will use the gene_id objects we made above to index and separate them.

num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_gene_ids])

# mrna promoter overlaps
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_gene_ids])



# Finding overlaps with gene_bodies (will take a few minutes again)
# Note this takes 10-15 min
genebody_peak_counts <- count_peaks_per_feature(mrna_lncrna_genes, 
                                                filtered_consensus_list, 
                                                type = "counts")

# Now let's extract the overlaps the same way we did for promoters above

# All gene bodies
num_peaks_df$peaks_overlapping_genebody <- rowSums(genebody_peak_counts)

# lncRNA gene bodies 
num_peaks_df$peaks_overlapping_lncrna_genebody <- rowSums(genebody_peak_counts[,lncrna_gene_ids])

# mRNA gene bodies
num_peaks_df$peaks_overlapping_mrna_genebody <- rowSums(genebody_peak_counts[,mrna_gene_ids])

# let's take a look.
# cool lots of info let's save:
write_csv(results/num_peaks_df, "num_peaks_df.csv")
```

```{r TF factor dataset}
# Downloading TF annotation data set
# this is a handy lesson of how to download any file in R
url <- "https://www.cell.com/cms/10.1016/j.cell.2018.01.029/attachment/ede37821-fd6f-41b7-9a0e-9d5410855ae6/mmc2.xlsx"
destination_for_url <- "results/TF_annotations.xlsx"

# to download we can use download.file
# download.file(url, destination_for_url)

# The above is a handy bit of code for downloading data into R directly,but BASH works to


# reading in DBP annotations as transcription factor

#redx1::read_excel to import
human_tfs <- readxl::read_excel("results/TF_annotations.xlsx",
                                sheet = 2, skip = 1)

human_tfs_full <- human_tfs
names(human_tfs_full)[4] <- "tf"

# let's rename the 4th column to indicate if it is a TF.
names(human_tfs)[4] <- "is_tf"

# merging annotation file to num_peaks_df
# first let's filter and grab the first 4 columns that match DBPs in num_peaks_df
human_tfs <- human_tfs[tolower(human_tfs$Name) %in% tolower(num_peaks_df$dbp), 1:4]

# adding new column names
names(human_tfs) <- c("ensembl_id",
                      "dbp",
                      "dbd",
                      "tf")

# Now comes the actual merge. Let's look at it first
num_peaks_df <- merge(num_peaks_df, human_tfs, all.x = T)

# Let's check how many NAs -- we should have some missing values.
dim(num_peaks_df[is.na(num_peaks_df$tf),])

# Ok looks good let's write it out:
num_peaks_df <- num_peaks_df[,1:12]
write_csv(num_peaks_df, "results/num_peaks_df.csv")

# Occurence matrix of overlaps
# note similar as above but using type = occurnece
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_list, 
                                               type = "occurrence")

# Let's double check that all lncrna & mrna genes are accounted for:
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# Great we will use this quite a bit moving forward so let's write it out! 
write.table(promoter_peak_occurence, "results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# First make sure promoter_peak_occurrence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))


# Now let's use the 'data.frame()' fucntion. Set up a bunch of colnames and populate them.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "chr" = lncrna_mrna_promoters@seqnames,   
                                "1kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))


# Let's write out this data frame -- all this code & run time is now encapsulated in one .csv :)
write_csv(peak_occurence_df, "results/peak_occurence_dataframe.csv")

save(filtered_consensus_list, gencode_genes, lncrna_gene_ids, mrna_gene_ids, num_peaks_df, peak_occurence_df, lncrna_mrna_promoters, mrna_lncrna_genes, human_tfs_full, file = "results/peak_features.RData")

# awesome - now we never have to load in consensus peaks again :)
# erase environment and reload!

#load("results/peak_features.RData", verbose = T)
# WOW that is super handy !!

```

```{r}

# Filter out any chip data less 1,000 peaks == filtered consensus peaks


# How does peak number and genome coverage compare
ggplot(num_peaks_df, aes(y = total_peak_length, x = num_peaks)) +
  geom_point() +
  # Adding a generative additive model 
  #geom_smooth(method = 'gam', formula = y ~ s(x, bs = "cs")) +
  # this adds the statistics from the gam to the figure
  #stat_cor() +
  geom_smooth(method = "lm") +
  # this is just to make it look nice.
  scale_x_continuous(expand = c(0,0)) +
  # title
  ggtitle("Num_Peaks_vs_Genome_Coverage") + 
  # labeling axes
  xlab(expression('Peak Number')) +
  ylab(expression('Genome Coverage')) 
ggsave("figures/peak_number_vs_coverage.pdf")

# What is the distribution of promoter overlaps versus gene-bodies (hint hist)
ggplot() +
  geom_histogram(aes(x = num_peaks_df$peaks_overlapping_promoters, 
                     fill = "peaks_overlapping_promoters"), alpha = 0.5) +
  geom_histogram(aes(x = num_peaks_df$peaks_overlapping_genebody, 
                     fill = "peaks_overlapping_genebody"), alpha = 0.5) +
  # title
  ggtitle("Peak Distribution") + 
  # labeling axes
  xlab(expression('Number of Peak Overlaps')) +
  ylab(expression('Frequency')) 
ggsave("figures/peak_distribution.pdf")

# Make a list of genes that are "super binders" 

# Is there a type of gene ontology associated with them versus the others?

# Is there a difference in mRNA and lncRNA promoter overlaps?

# Do lncRNAs also have super-binding promoters?

# How many of these proteins are TFs? What is the most represented type of DBD?


```
