---
title: "02_clustering"
author: "kurt"
date: "4/24/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
library(ggdendro)
library(pheatmap)

setwd("/scratch/Shares/rinnclass/CLASS_2023/kurt/CLASS_2023/Final_Project/02_clustering")

# create folders
if (dir.exists("figures") == F){
  dir.create("figures")}
```


```{r loading in promoter_peak_occurence_df}
# load in file
promoter_peak_occurence_matrix <- read.table("../01_peaks/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")
# Converting to a matrix format for correlation analysis
promoter_peak_occurence_matrix <- as.matrix(promoter_peak_occurence_matrix)
# This is a very large matrix of 1 and 0 if a DBP is bound to a promoter.
```

# calculating distance matrix
Let's get started by finding the 'distance' between each promtoer vector.
We will make a distance matrix and then use that for clustering.
```{r calculating distance matrix}

# Distance matrix for each pairwise vector comparison for the whole matrix
peak_occurence_dist <- dist(promoter_peak_occurence_matrix, method = "binary")
```

# Clustering distance matrix

Now that we have that, we can do hierarchical clustering. In brief, this will cluster 
vectors that have small distances (high correlation), and will iterate that process until
everything is in the same cluster.

```{R hclust of distance matrix}
#cluster
bin_hier <- hclust(peak_occurence_dist, method = "complete")

pdf("figures/dbp_hclust_dendro.pdf", height = 12, width = 70)
plot(bin_hier)
dev.off()

plot(bin_hier)

```


# ggdendro function for clustwering
Nice so we now have object "bin_hier" that will contain all the cross correlation values across all the samples. 

Now let's use GGDENDRO (dendrogram) package that will plot the branch lengths that indicate how similar two samples are 

```{r ggdendro gram}

# This is a lot of ggplot -- on purpose
# please read through what is happening and parameter usage

 ggdendro::ggdendrogram(bin_hier, rotate = FALSE,  size = 3, 
                       theme_dendro = TRUE) +
   # 90 degree rotation to right
   coord_flip() +
   scale_y_continuous() +
   # adds label
   scale_x_continuous(position = "top") +
   # subsection to labels in order of clustering
   # ? seq_along
   scale_x_continuous(breaks = seq_along(bin_hier$labels[bin_hier$order]),
                      
                      # adding labels that are in the order 'column'
             labels = bin_hier$labels[bin_hier$order], position = "top",
             expand = c(0,0)) +
   theme(axis.text.x = element_text(angle = 90, hjust  = 1)) + 
   theme(axis.text.y = element_text(angle = 0,hjust = 1)) +
   scale_y_reverse(expand = c(0.01, 0)) +
   theme(
     plot.background = element_blank(),
     panel.grid.major = element_blank(),
   panel.grid.minor = element_blank(),
     panel.border = element_blank()
   )

# Nice let's save this using 'ggsave': since we used ggplot

ggsave("figures/ggdendro_plot.png", height = 50, width = 12, limitsize = F)

```


# Clustering of lncRNA and mRNA seperately

# lncRNA promoters ggdendro
```{r lncRNA promoter clustering}

# lncrna and mrna promoters
lncrna_mrna_promoters <- rtracklayer::import("../01_peaks/results/gene_annotations/lncrna_mrna_promoters.gtf")

# lncrna_promoters
lncrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "lncRNA"]

# Now we will use indexing to separate peak_occurrence_matrix in lncRNA and mRNA.
# note we are indexing and using indexing in indexing to get what we want.
lncrna_peak_occurence <- promoter_peak_occurence_matrix[,lncrna_promoters$gene_id]

# we do the clutering the same as above or we can combine the dist and hclust:
bin_hier_lncrna <- hclust(dist(lncrna_peak_occurence, method = "binary"))

# Now plot with ggdendro
ggdendro::ggdendrogram(bin_hier_lncrna, rotate = T,  size = 3)
 
# Now let's save this figure
ggsave("figures/lncrna_hclust_binary_dist.png", height = 49, width = 6)


```

# mRNA promter ggdendro
Now for mRNA

```{R mRNA promoter clustering}

# mrna promoters
mrna_promoters <- lncrna_mrna_promoters[lncrna_mrna_promoters$gene_type == "protein_coding"]

# same strategy used very often:
# indexing into larger file and filtering to a file of stuff you want.
mrna_peak_occurence <- promoter_peak_occurence_matrix[,mrna_promoters$gene_id]

# getting the distance matrix for only mRNA promoters  
bin_hier_mrna <- hclust(dist(mrna_peak_occurence, method = "binary"))
 
# plotting with ggdendro
ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)

# saving
ggsave("figures/mrna_hclust_binary_dist.png", height = 49, width = 6)


```

# Characterizing and Clustering "high binder promoters"
Let's make a heatmap of the promoters -- since it will take a long long time to plot we'll just plot the "super promoters"
```{r clustering high-binding promoters}
# first we filter promoter_peak_occurrence to just those that have high numbers of dbps
# let's, find the promoters that bind to many of our dbps
# a bit of trial and error but 5 of 7 dbps works for me.
hist(colSums(promoter_peak_occurence_matrix))

# define 200 as high binders from histogram
high_binders <- promoter_peak_occurence_matrix[,colSums(promoter_peak_occurence_matrix) > 200]

# let's find out how many high binder promoters there are
ncol(high_binders)

```


```{r Get list of of promoters a dbp binds}
# Step 1: make a dataframe of gene ids and gene_names
DBPs_on_promoter <- bind_rows(lncrna_promoters %>% as.data.frame(),
                              mrna_promoters %>% as.data.frame()) %>%
                              dplyr::select(gene_id, gene_name)

# Step 2: Make a table to look up which DBPs are on a given promoter
promoter_dbps <- promoter_peak_occurence_matrix %>%
  as.data.frame() %>%
  rownames_to_column("dbp") %>%
  pivot_longer(2:ncol(.), names_to = "gene_id", values_to = "occurrence") %>%
# This now a very long matrix - let's get rid of all zero values
  filter(occurrence == 1) %>%
  dplyr::select(-occurrence) %>%
  left_join(DBPs_on_promoter)

# let's write this out it's a very handy DF!
# note saving in 12 where we read in this type of data
write.csv(promoter_dbps, "../01_peaks/results/dbps_on_a_promoter.csv")
# This is a very large file but very useful you want to find which DBP is on which promoter

```

# Finding which DBPs are on which promoter
Cool now we can select any gene and see what DBPs are bound?

```{R DBPs on a given promoter}

# In the future if we want to filter on genes we can just read this back in.
promoter_dbps <- read.csv("../01_peaks/results/dbps_on_a_promoter.csv")

# let's see what's on FIRRE
firre_promoter <- promoter_dbps %>%
  filter(gene_name == "FIRRE")

# seeing what's there :
firre_promoter
# How many DBPs bound
nrow(firre_promoter)

# GAPDH
GAPDH_promoter <- promoter_dbps %>%
  filter(gene_name == "GAPDH")

# seeing what's there :
GAPDH_promoter
# How many DBPs bound
nrow(GAPDH_promoter)

# trying to find a small binder :)
XIST_promoter <- promoter_dbps %>%
  filter(gene_name == "XIST")

# seeing what's there :
XIST_promoter
# How many DBPs bound
nrow(XIST_promoter)

# Interesting, intuition was correct considering
# this is a male cell line :)
```

