<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="kurt" />


<title>Final_class_project</title>

<script src="Winkler_Final_Project_files/header-attrs-2.20/header-attrs.js"></script>
<script src="Winkler_Final_Project_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Winkler_Final_Project_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="Winkler_Final_Project_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Winkler_Final_Project_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Winkler_Final_Project_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Winkler_Final_Project_files/navigation-1.1/tabsets.js"></script>
<link href="Winkler_Final_Project_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Winkler_Final_Project_files/highlightjs-9.12.0/highlight.js"></script>
<link href="Winkler_Final_Project_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="Winkler_Final_Project_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Final_class_project</h1>
<h4 class="author">kurt</h4>
<h4 class="date">4/24/2023</h4>

</div>


<div id="your-mission-should-you-choose-to-accept-it" class="section level1">
<h1>Your mission should you choose to accept it</h1>
<p>There is an enormous amount of data in ENCODE that needs to be analyzed. This includes 480 unique DBP ChIPseq datasets. To facilitate your mission we provide you 1,066 peak files that have been run through the NF_CORE chipseq pipeline.</p>
<pre class="r"><code>load(&quot;01_peaks/results/peak_features.RData&quot;, verbose = T)
load(&quot;01_peaks/results/peak_features2.RData&quot;, verbose = T)
load(&quot;04_RNAseq/results/nuc_cyto.RData&quot;, verbose = T)</code></pre>
</div>
<div id="first-i-analyze-the-large-data-for-general-properties" class="section level1">
<h1>First I analyze the large data for general properties</h1>
<pre class="r"><code># Any chip data less 1,000 peaks was filtered out

# How does peak number and genome coverage compare
knitr::include_graphics(&quot;01_peaks/figures/peak_number_vs_coverage.pdf&quot;)</code></pre>
<p><embed src="01_peaks/figures/peak_number_vs_coverage.pdf" /><!-- --></p>
<pre class="r"><code># What is the distribution of promoter overlaps versus gene-bodies (hint hist)
knitr::include_graphics(&quot;01_peaks/figures/peak_distribution.pdf&quot;)</code></pre>
<p><embed src="01_peaks/figures/peak_distribution.pdf" /><!-- --></p>
<pre class="r"><code># Make a list of genes that are &quot;super binders&quot; 
super_binders &lt;- peak_occurence_df[peak_occurence_df$number_of_dbp &gt; 200,]

# Is there a type of gene ontology associated with them versus the others?
cat(&quot;Most of the super binders are associated with mRNA&quot;)</code></pre>
<pre><code>## Most of the super binders are associated with mRNA</code></pre>
<pre class="r"><code># Is there a difference in mRNA and lncRNA promoter overlaps?
cat(paste0(&quot;Number of lncRNA super binders: &quot;, 
           nrow(super_binders %&gt;% filter(gene_type==&quot;lncRNA&quot;)), 
           &quot;\n&quot;,
           &quot;Number of mRNA super binders: &quot;, 
           nrow(super_binders %&gt;% filter(gene_type==&quot;protein_coding&quot;))))</code></pre>
<pre><code>## Number of lncRNA super binders: 2512
## Number of mRNA super binders: 9177</code></pre>
<pre class="r"><code># Do lncRNAs also have super-binding promoters?
cat(&quot;Yes, about 21.5% of the super-binding promoters are associated with lncRNAs&quot;)</code></pre>
<pre><code>## Yes, about 21.5% of the super-binding promoters are associated with lncRNAs</code></pre>
<pre class="r"><code># How many of these proteins are TFs? What is the most represented type of DBD?
supers &lt;- human_tfs_full %&gt;%
  filter(ID %in% sub(&quot;\\..*&quot;, &quot;&quot;, super_binders$gene_id),
         tf == &quot;Yes&quot;)

cat(paste0(&quot;Number of TFs: &quot;, 
           nrow(supers %&gt;% filter(tf==&quot;Yes&quot;)), 
           &quot;\n&quot;,
           &quot;Most represented type of DBD: &quot;, 
           names(which.max(table(supers$DBD)))))</code></pre>
<pre><code>## Number of TFs: 811
## Most represented type of DBD: C2H2 ZF</code></pre>
</div>
<div id="results" class="section level1">
<h1>Results</h1>
<p>Peak number and genome coverage have a linear relationship</p>
<p>There is a larger distribution of peaks overlapping gene bodies compared to promoters</p>
<p>Super-binders are more associated with mRNAs</p>
<p>I found 9,177 mRNA superbinding promoters and 2512 lncRNA super-promoters</p>
<p>21.5% of super-binding promoters are associated with lncRNAs</p>
<p>811 of the proteins are TFs and the most represented DBD is C2H2 ZF</p>
</div>
<div id="clustering---how-similar-our-genes-are-to-each-other" class="section level1">
<h1>Clustering - how similar our genes are to each other</h1>
<pre class="r"><code># What genes make sense to cluster together create an image (hint Pol II makes RNA)
cat(&quot;Genes that make sense to cluster together are ones associated with either lncRNA or mRNA&quot;, 
    &quot;as they are more likely to have similar functions. It also makes sense to cluster genes&quot;, 
    &quot;that may be in the same pathway to see if they are related to each other.&quot;)</code></pre>
<pre><code>## Genes that make sense to cluster together are ones associated with either lncRNA or mRNA as they are more likely to have similar functions. It also makes sense to cluster genes that may be in the same pathway to see if they are related to each other.</code></pre>
<pre class="r"><code># Find a cluster of genes your interested in and can figure out what those genes do -- are there unknown genes in there too? If so maybe one could hypothesize they have a similar function to the known genes (Hint ZNFs)
knitr::include_graphics(&quot;02_clustering/figures/dbp_hclust_dendro.pdf&quot;)</code></pre>
<p><embed src="02_clustering/figures/dbp_hclust_dendro.pdf" /><!-- --></p>
<pre class="r"><code>cat(paste0(&quot;Gene Cluster: NRL, ZNF280B&quot;,
           &quot;\n&quot;,
           &quot;ZNF280B appears to be similar to gene NRL and may be involved in photorecptor development/function&quot;))</code></pre>
<pre><code>## Gene Cluster: NRL, ZNF280B
## ZNF280B appears to be similar to gene NRL and may be involved in photorecptor development/function</code></pre>
<pre class="r"><code># if we cluster by lncRNA and mRNA separately what are some similarities and differences?
knitr::include_graphics(&quot;02_clustering/figures/lncrna_hclust_binary_dist.pdf&quot;)</code></pre>
<p><embed src="02_clustering/figures/lncrna_hclust_binary_dist.pdf" /><!-- --></p>
<pre class="r"><code>knitr::include_graphics(&quot;02_clustering/figures/mrna_hclust_binary_dist.pdf&quot;)</code></pre>
<p><embed src="02_clustering/figures/mrna_hclust_binary_dist.pdf" /><!-- --></p>
<pre class="r"><code>cat(paste0(&quot;Similarities: similar amount of genes closely clustered, some genes tightly clustered in both (FOXA1/2)&quot;,
           &quot;\n&quot;,
           &quot;Differences: different overall shape of plot, some genes clustered differently (YEATS2)&quot;))</code></pre>
<pre><code>## Similarities: similar amount of genes closely clustered, some genes tightly clustered in both (FOXA1/2)
## Differences: different overall shape of plot, some genes clustered differently (YEATS2)</code></pre>
</div>
<div id="results-1" class="section level1">
<h1>Results</h1>
<p>ZNF280B appears to be similar to gene NRL and may be involved in photoreceptor development/function</p>
</div>
<div id="metaplots" class="section level1">
<h1>Metaplots</h1>
<pre class="r"><code># Let&#39;s look at the metaplot for all DBPs on lncRNA and mRNA promoters seperately (hint facet wrap).
knitr::include_graphics(&quot;03_metaplots/figures/lncrna_mrna_metaplot.png&quot;)</code></pre>
<p><img src="03_metaplots/figures/lncrna_mrna_metaplot.png" width="1567" /></p>
<pre class="r"><code># Which genes seem to have a difference in where they bind on promoters between lncRNA and mRNA promoters
cat(&quot;H3K4me1, H3K4me2, H3K79me2, H4K20me1, EZH2&quot;)</code></pre>
<pre><code>## H3K4me1, H3K4me2, H3K79me2, H4K20me1, EZH2</code></pre>
<pre class="r"><code># Make a metaplot of DBPS only on Super-binders versus regular promoters
knitr::include_graphics(&quot;03_metaplots/figures/high_low_metaplot.png&quot;)</code></pre>
<p><img src="03_metaplots/figures/high_low_metaplot.png" width="1563" /></p>
<pre class="r"><code>cat(&quot;Super-binders are more distributed&quot;)</code></pre>
<pre><code>## Super-binders are more distributed</code></pre>
</div>
<div id="results-2" class="section level1">
<h1>Results</h1>
<p>H3K4me1, H3K4me2, H3K79me2, H4K20me1, and EZH2 bind differently on lncRNA and mRNA promoters</p>
<p>DBPs on super-binders tend to be more distributed than on regular promoters</p>
</div>
<div id="rnaseq-expression" class="section level1">
<h1>RNAseq expression</h1>
<pre class="r"><code># What is the relationship between number of DBPS bound on a promoter versus RNA output (hint TPM)
knitr::include_graphics(&quot;04_RNAseq/figures/Expression_vs_DBP_promoter_binding.pdf&quot;)</code></pre>
<p><embed src="04_RNAseq/figures/Expression_vs_DBP_promoter_binding.pdf" /><!-- --></p>
<pre class="r"><code># Let&#39;s make a heatmap of genes that are variable across samples 
knitr::include_graphics(&quot;04_RNAseq/figures/all_vs_total_genes_heatmap.pdf&quot;)</code></pre>
<p><embed src="04_RNAseq/figures/all_vs_total_genes_heatmap.pdf" /><!-- --></p>
<pre class="r"><code># Which subfraction of the cell has the highest expression.
cat(&quot;The nuclear subfraction of the cell has the highest expression&quot;)</code></pre>
<pre><code>## The nuclear subfraction of the cell has the highest expression</code></pre>
<pre class="r"><code># Let&#39;s make a heatmap of nuclear versus cytoplasmic expression
knitr::include_graphics(&quot;04_RNAseq/figures/nuc_vs_cyto_genes_heatmap.pdf&quot;)</code></pre>
<p><embed src="04_RNAseq/figures/nuc_vs_cyto_genes_heatmap.pdf" /><!-- --></p>
<pre class="r"><code># How many lncRNA and mRNA genes are sig in nuclear or cyto
cat(paste0(&quot;Significant lncRNA Nuclear Genes: &quot;, nrow(lncrna_nuc_sig), &quot;\n&quot;,
           &quot;Significant mRNA Nuclear Genes: &quot;, nrow(mrna_nuc_sig), &quot;\n&quot;,
           &quot;Significant lncRNA Cytosolic Genes: &quot;, nrow(lncrna_cyto_sig), &quot;\n&quot;,
           &quot;Significant mRNA Cytosolic Genes: &quot;, nrow(mrna_cyto_sig)))</code></pre>
<pre><code>## Significant lncRNA Nuclear Genes: 2115
## Significant mRNA Nuclear Genes: 2559
## Significant lncRNA Cytosolic Genes: 917
## Significant mRNA Cytosolic Genes: 3499</code></pre>
<pre class="r"><code># What is/are the most nuclear mRNA(s) -- is there a type of gene ontology associated with them?
nuc_mrnas &lt;- nuc_cyto_sig_genes %&gt;%
  filter(gene_id %in% mrna_nuc_sig$gene_id)
nuc_mrnas[order(nuc_mrnas$log2FoldChange)[1:10],]$gene_name</code></pre>
<pre><code>##  [1] &quot;MT-CO2&quot;  &quot;MT-ND2&quot;  &quot;MT-ND1&quot;  &quot;MT-ND6&quot;  &quot;MT-ND3&quot;  &quot;MT-ND5&quot;  &quot;MT-ATP8&quot;
##  [8] &quot;MT-ND6&quot;  &quot;MT-CYB&quot;  &quot;MT-CO1&quot;</code></pre>
<pre class="r"><code>print(&quot;The most nuclear mRNAs are mitochondrial genes&quot;)</code></pre>
<pre><code>## [1] &quot;The most nuclear mRNAs are mitochondrial genes&quot;</code></pre>
<pre class="r"><code># If we zoom in on high binding promoters (&gt; 200 DBPs) are there any that don&#39;t have any expression?
cat(&quot;Yes,&quot;, nrow(promoter_features_df %&gt;% 
       filter(hepg2_reservoir == 1)), &quot;reservoirs were found&quot;)</code></pre>
<pre><code>## Yes, 168 reservoirs were found</code></pre>
</div>
<div id="results-3" class="section level1">
<h1>Results</h1>
<p>There is a linear relationship between RNA expression and DBPs bound</p>
<p>The nuclear subfraction of the cell has the highest expression</p>
<p>The most nuclear mRNAs are mitochondrial genes</p>
<p>in 04_RNAseq/results/reservoirs.csv I have all the super binders that donâ€™t express RNA</p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
